programs/mango-v4/src/instructions/openbook_v2_place_order.rs

```diff
@@ -0,0 +1,538 @@
+        require!(
+            account
+                .fixed
+                .is_owner_or_delegate(ctx.accounts.authority.key()),
+            MangoError::SomeError
+        );
+
+        // Validate open_orders #2
+        require!(
+            account
+                .openbook_v2_orders(openbook_market.market_index)?
+                .open_orders
+                == ctx.accounts.open_orders.key(),
+            MangoError::SomeError
+        );
+    }
+    // Validate bank and vault #3
+    let group_key = ctx.accounts.group.key();
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    let retriever = new_fixed_order_account_retriever(ctx.remaining_accounts, &account.borrow())?;
+
+    let (_, _, payer_active_index) = account.ensure_token_position(receiver_token_index)?;
+    let (_, _, receiver_active_index) = account.ensure_token_position(receiver_token_index)?;
+
+    let (payer_bank, payer_bank_oracle) =
+        retriever.bank_and_oracle(&group_key, payer_active_index, payer_token_index)?;
+    let (receiver_bank, receiver_bank_oracle) =
+        retriever.bank_and_oracle(&group_key, receiver_active_index, receiver_token_index)?;
+
+    require_keys_eq!(payer_bank.vault, ctx.accounts.payer_vault.key());
+    require_eq!(payer_bank.token_index, payer_token_index);
+    require_eq!(receiver_bank.token_index, receiver_token_index);
+
+    //
+    // Pre-health computation
+    //
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    let mut health_cache = new_health_cache(&account.borrow(), &retriever, now_ts)
+        .context("pre-withdraw init health")?;
+    let pre_health_opt = if !account.fixed.is_in_health_region() {
+        let pre_init_health = account.check_health_pre(&health_cache)?;
+        Some(pre_init_health)
+    } else {
+        None
+    };
+
+    drop(retriever);
+
+    // No version check required, bank writable from v1
+
+    //
+    // Before-order tracking
+    //
+    let openbook_market_external = ctx.accounts.openbook_v2_market_external.load()?;
+    let before_vault = ctx.accounts.payer_vault.amount;
+
+    let before_oo_free_slots;
+    let before_had_bids;
+    let before_had_asks;
+    let before_oo = {
+        let open_orders = ctx.accounts.open_orders.load()?;
+        before_oo_free_slots = MAX_OPEN_ORDERS - open_orders.all_orders_in_use().count();
+        before_had_bids = open_orders.position.bids_base_lots != 0;
+        before_had_asks = open_orders.position.asks_base_lots != 0;
+        OpenOrdersSlim::from_oo_v2(
+            &open_orders,
+            openbook_market_external.base_lot_size,
+            openbook_market_external.quote_lot_size,
+        )
+    };
+
+    // Provide a readable error message in case the vault doesn't have enough tokens
+    let base_lot_size;
+    let quote_lot_size;
+    {
+        base_lot_size = openbook_market_external.base_lot_size;
+        quote_lot_size = openbook_market_external.quote_lot_size;
```

Comment:
minor: move the base_lot_size assignments into the `let` declarations. The whole `base/quote_lot_size` constants could be set after openbook_market_external was loaded and then possibly used instead of the full `openbook_market_external.base_lot_size` (right now some places use one and some the other)